File: benchmarks/lang/cpp/matmul.cpp
// Dense matrix multiply, same algorithm as the Lua/Tzopilotl ports.
#include <cstdio>
#include <cmath>
#include <vector>

static const int N = 500;
static const double FN = 500.0;

static std::vector<std::vector<double>> init_matrix(double phase) {
    std::vector<std::vector<double>> rows(N, std::vector<double>(N));
    for (int i = 0; i < N; ++i)
        for (int j = 0; j < N; ++j)
            rows[i][j] = std::sin(i * FN + j + phase);
    return rows;
}

int main() {
    auto A = init_matrix(0.0);
    auto B = init_matrix(0.5);

    std::vector<std::vector<double>> C(N, std::vector<double>(N));
    for (int i = 0; i < N; ++i) {
        auto const& Ai = A[i];
        for (int j = 0; j < N; ++j) {
            double s = 0.0;
            for (int k = 0; k < N; ++k) s += Ai[k] * B[k][j];
            C[i][j] = s;
        }
    }

    double trace = 0.0;
    for (int i = 0; i < N; ++i) trace += C[i][i];
    std::printf("%.15g\n", trace);
}

File: benchmarks/lang/lua/matmul.lua
-- Dense matrix multiply, same algorithm as the Tzopilotl port.
local N = 500
local FN = 500.0
local sin = math.sin

local function init_matrix(phase)
    local rows = {}
    for i = 0, N - 1 do
        local row = {}
        for j = 0, N - 1 do
            row[j] = sin(i * FN + j + phase)
        end
        rows[i] = row
    end
    return rows
end

local A = init_matrix(0.0)
local B = init_matrix(0.5)

local C = {}
for i = 0, N - 1 do
    local Ci = {}
    local Ai = A[i]
    for j = 0, N - 1 do
        local s = 0.0
        for k = 0, N - 1 do
            s = s + Ai[k] * B[k][j]
        end
        Ci[j] = s
    end
    C[i] = Ci
end

local trace = 0.0
for i = 0, N - 1 do trace = trace + C[i][i] end
print(trace)
